src/main/java/com/antelif/library/domain/service/validation/TransactionValidationService.java

```diff
@@ -57,6 +60,31 @@ public static void validateUpdate(
     validateBookCopiesToReturnAreLent(transactions, bookCopyIdsReturned);
   }
 
+  /**
+   * Throws an exception if the transaction provided to cancel has been finalized.
+   *
+   * @param transaction the transaction to cancel.
+   */
+  public static void validateCancel(TransactionEntity transaction) {
+    validateTransactionIsFinalized(transaction);
+    validateTransactionIsPartiallyUpdated(transaction);
+  }
+
+  private static void validateTransactionIsFinalized(TransactionEntity transaction) {
+    if (transaction.getStatus().equals(FINALIZED)) {
+      throw new UnsuccessfulTransactionException(CANNOT_CANCEL_TRANSACTION);
+    }
+  }
+
+  private static void validateTransactionIsPartiallyUpdated(TransactionEntity transaction) {
+    if (transaction.getTransactionItems().stream()
+        .map(TransactionItemEntity::getBookCopy)
+        .map(BookCopyEntity::getStatus)
+        .anyMatch(status -> status.equals(AVAILABLE))) {
+      throw new UnsuccessfulTransactionException(CANNOT_CANCEL_TRANSACTION);
```

Comment:
This error does not refer to a finalized transaction. Maybe create a new error to display for this one.